include/seqan3/argument_parser/auxiliary.hpp

```diff
@@ -15,12 +15,211 @@
+{}; // forward
+
+//!\cond
+template <typename t>
+struct argument_parsing<t const> : argument_parsing<t>
+{};
+
+template <typename t>
+struct argument_parsing<t &> : argument_parsing<t>
+{};
+
+template <typename t>
+struct argument_parsing<t const &> : argument_parsing<t>
+{};
+//!\endcond
+
+} // seqan3::custom
+
+namespace seqan3::detail::adl_only
+{
+
+//!\brief Poison-pill overload to prevent non-ADL forms of unqualified lookup.
+template <typename t>
+std::unordered_map<std::string_view, t> enumeration_names(t v) = delete;
+
+/*!\brief Functor definition for seqan3::enumeration_names.
+ *
+ * This struct is templetised because we need the original option_t next to s_option_t that may be wrapped
+ * std::type_identity to be default constructible. We need option_t to be default constructible because the
+ * respective CPO should be callable without a function parameter.
+ */
+template <typename option_t>
+struct enumeration_names_fn
+{
+private:
+    //!\brief `option_t` with cvref removed and possibly wrapped in std::type_identity.
+    using s_option_t = std::conditional_t<std::is_nothrow_default_constructible_v<remove_cvref_t<option_t>> &&
+                                        seqan3::is_constexpr_default_constructible_v<remove_cvref_t<option_t>>,
+                                        remove_cvref_t<option_t>,
+                                        std::type_identity<option_t>>;
+
+    SEQAN3_CPO_IMPL(1, (deferred_type_t<seqan3::custom::argument_parsing<option_t>, decltype(v)>::enumeration_names))
+    SEQAN3_CPO_IMPL(0, (enumeration_names(v))) // ADL
+
+public:
+    //!\brief Operator definition.
+    template <typename dummy = int> // need to make this a template to enforce deferred initialisation
+    //!\cond
+        requires requires
+        {
+            { impl(priority_tag<1>{}, s_option_t{}, dummy{}) } -> std::unordered_map<std::string_view,
+                                                                                     remove_cvref_t<option_t>>;
+        }
+    //!\endcond
+    auto operator()() const noexcept
+    {
+        return impl(priority_tag<1>{}, s_option_t{});
+    }
+};
+
+} // namespace seqan3::detail::adl_only
 
 namespace seqan3
 {
 
+/*!\name Customisation Points
+ * \{
+ */
+
+/*!\brief Return a conversion map from std::string_view to option_type.
+ * \tparam your_type Type of the value to retrieve the conversion map for.
+ * \param value The value is not used, just its type.
+ * \returns A std::unordered_map<std::string_view, your_type> that maps a string identifier to a value of your_type.
+ * \ingroup argument_parser
+ * \details
+ *
+ * This is a function object. Invoke it with the parameter(s) specified above.
+ *
+ * It acts as a wrapper and looks for two possible implementations (in this order):
+ *
+ *   1. A static member `enumeration_names` in `seqan3::custom::argument_parsing<your_type>` that is of type
+ *      `std::unordered_map<std::string_view, your_type>>`.
```

Comment:
Can/should this be const? Is that enforced or unimportant?

Reply:
Hm I guess it could but I don't think it is important? I don't know...

Reply:
Hm, I think it should be const...

Reply:
I think we should relax these constraints to just be an "associative range", but that can be done later.

Reply:
But I cannot enforce in the concept that the impl with a static member must return a const member while the free function returns a copy?

Reply:
The concept doesn't know about the impls. The concept just accesses the trait / CPO. And that should always be const.